### PR TITLE
[Bugfix/refinement] ota-proxy timeout issue

### PR DESCRIPTION
**Pull request by Bodong-Yang**
## Introduction
This PR fixes the issue caused by unwanted timeout implemented in previous PR for ota-proxy, and optimizes some implementation.

## Changes
1. fix and refine `OngoingCacheTracker`, resolve the unwanted timeout issue, resolve the edge condition that subscriber subscribes on closed tracker,
2. move more configuration to `ota_proxy.config` instead of hard-coded,
3. fix and refine the `RemoteOTAFile`, simplify the background_cache_writing logic and cache commit logic,
4. refine `OTACache.retrieve_file`,
5. deps: bump uvicorn==0.20.0, aiofiles==22.1.0, and pin aiohttp==3.8.1,
6. optimize ota_proxy configuration according to test result,
7. test files updates,
8. other minor updates.

## Optimization comes with this PR
1. `MAX_DOWNLOAD_THREAD` reduces from 8 to 3, as the benefit is not significant,
2. use `uvloop` for `ota-proxy`, which results in around 10%~ or more performance improved,
3. reduce `CHUNK_SIZE` to 1MB, tests reveal that this setting has better performance,
4. reduce `STREAMING_BACKOFF_FACTOR` to 0.001 to reduce blocking

## Ticket
[T4PB-24098](https://tier4.atlassian.net/browse/T4PB-24098)

## Test
- [x] single ECU: real network
- [x] single ECU: local test
- [x] multiple ECUs: local test
- [x] multiple ECUs: real network
- [x] test files

### CODE REVIEW

**Comment by Bodong-Yang** [comment]
Local test environment now use `tc` and **VirtualBox storage IO limitation** to simulate ota-proxy running on a SATA SSD backed device with LTE-like network environment.

### Network environment simulation with `tc`
```
tc qdisc add dev vboxnet0 root netem delay 43ms 5ms 25% loss 0.05% 25% limit 6k rate 120Mbit
```
This will apply a average 43ms delay with 5ms jitter, 0.05% package loss rate to the virtualbox network adapter, and limit the bandwidth to 120Mbit. This setting tries to simulate the LTE network environment.

### Limit the IO bandwidth of autoware ECU VM
```
# define limitation group for the VM
VBoxManage bandwidthctl "ota-vm" add LimitDynamic --type disk --limit 400M 
# apply limitation group to disk
VBoxManage storageattach "ota-vm" --storagectl "SATA" --port 0  --device 0 --type hdd --medium <path_to_vm>/ota-vm-disk001.vdi --bandwidthgroup LimitDynamic
# change the limitation online
VBoxManage bandwidthctl "ota-vm" set LimitDynamic --limit 500M
``` 
This will simulate the IO bandwidth as SATA SSD, but the latency is not considered.

**Comment by Bodong-Yang** [comment]
Local multiple ECUs test is simulated by running one ota-proxy instance on VM, and launch 5 client on host machine.
For a ~5.9G ota-image, it takes 34mins for 5 clients to finish the downloading via ota-proxy.

**Comment by Bodong-Yang** [comment]
Mutiple ECUs with real network(100M fiber network) is ongoing.

**Comment by Bodong-Yang** [comment]
# Test result
In real condition, it is unlike that otaclient needs to download the full ota-image, so in all the tests clients are configure to download only half of the ota-image.
The **client** here is a script that uses the `Downloader` from previous version of otaclient to directly download files from OTA image server via ota-proxy.

## Local test
### Test setup
1. one ota-proxy instance on VM, 5 clients on host machine
2. network: `tc` controlled network(43ms latency, 5ms jitter, 0.01% packet loss rate, 120Mbit bandwidth)
3. disk IO for ota-proxy VM: 500M
### Test result
1. Downloaded files: 147340, ~10G
2. Time cost: 2313s
3. Finished without errors

## Real network test
### Test setup
1. one ota-proxy instance on VM, 5 clients on host machine
2. network: 100Mbit fiber network
3. disk IO for ota-proxy VM: 500M
### Test result
1. Downloaded files: 149790, ~14G
2. Time cost: 3291s
3. Finish without errors

**Comment by Bodong-Yang** [comment]
TODO: test files update

**Comment by Bodong-Yang** [comment]
The downloading speed is related to the bandwidth and network latency.
In testing, a 40ms latency LTE can be 3~4times slower than a 6ms networking, it impacts more than the bandwidth because the downloading is hard to utilize full bandwidth when in high latency network.

**Comment by obi-t4** [comment]
@Bodong-Yang 

Thank you for the evaluation!

> a 40ms latency LTE can be 3~4times slower than a 6ms networking

`6ms networking` means `a 6ms latency network`?

**Comment by Bodong-Yang** [comment]
> @Bodong-Yang
> 
> Thank you for the evaluation!
> 
> > a 40ms latency LTE can be 3~4times slower than a 6ms networking
> 
> `6ms networking` means `a 6ms latency network`?

Yes, I use tc to apply 6ms on all traffics in the vboxnet0 to simulate VM using a 6ms latency network.

The `tc` setups applied to vboxnet0 are as follow:
1. to simulate LTE connection: 43ms latency, 5ms jitter, 0.01% packet loss rate
2. to simulate a 100Mbit fast fiber network: 6ms latency, no jitter, no packet loss

**Comment by obi-t4** [comment]
current main branch fails under this `43ms latency, 5ms jitter, 0.01% packet loss rate` condition and this PR succeeds under that condition?

**Comment by Bodong-Yang** [comment]
The reason why latency impacts so much is because most of the files in ota-image are quite small, less than 1MB, so the impact from the overhead of each file will overwhelming the impact from the file size.
In high latency network, downloading 100 files with 1MB size will be much much slower than download 1 file with 100MB size.

**Comment by Bodong-Yang** [comment]
> current main branch fails under this `43ms latency, 5ms jitter, 0.01% packet loss rate` condition and this PR succeeds under that condition?

Yes, the current PR fix succeeded all test cases, only the matter of time cost.

**Comment by Bodong-Yang** [comment]
Note that there is another strange thing happens, there is one edge condition for ota-proxy cache streaming for multiple request to same file that subscriber subscribes on a just closed tracker.
This edge condition is hard to trigger and can only result in one hash mismatched(retry one more time can resolve this issue and download the correct file). But in my local pytest running, this edge condition being triggered every time on the same file.
So I implement a simple try(directly retry at most 6 times) for the `Downloader` used in `test_ota_proxy_server.py`. I think it is OK as the real `Downloader` used by otaclient will also do the same retry on hash mismatch.

File: otaclient/ota_proxy/config.py [hunk 1]
```diff
@@ -45,8 +42,15 @@ class Config:
...
+    DB_THREAD_POOL_SIZE = 3
     TABLE_DEFINITION_VERSION = "v3"
-    TABLE_NAME: str = f"ota_cache_{TABLE_DEFINITION_VERSION}"
+    TABLE_NAME = f"ota_cache_{TABLE_DEFINITION_VERSION}"
+
+    # cache streaming behavior
+    STREAMING_WAIT_FOR_FIRST_BYTE = 3  # second
+    STREAMING_TIMEOUT = 10  # seconds
+    STREAMING_BACKOFF_FACTOR = 0.001  # second
+    STREAMING_WAIT_FOR_WRITER = 0  # second
```
Comment (obi-t4) [inline otaclient/ota_proxy/config.py:53]: It is better to set non-zero not to exhaust resource, I think.

**Review by obi-t4 — COMMENTED** [review]
I left some comments. But basically OK.

File: otaclient/app/configs.py [hunk 2]
```diff
@@ -94,7 +94,7 @@ class BaseConfig:
     DOWNLOAD_RETRY: int = 10
     DOWNLOAD_BACKOFF_MAX: int = 3  # seconds
     MAX_CONCURRENT_TASKS: int = 128
-    MAX_DOWNLOAD_THREAD: int = 8
+    MAX_DOWNLOAD_THREAD = 3
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/app/configs.py:97]: Emmm basically I found that 8 threads is too much and gain not so much improvement over only use 3 download threads. That is not related to bug fix though, if this change needed to be introduced separately, I can revert it can introduce it later.

File: otaclient/app/ota_client_stub.py [hunk 3]
```diff
@@ -86,6 +87,7 @@ async def _start_uvicorn(init_cache: bool, *, scrub_cache_event):
             server = uvicorn.Server(config)
             await server.serve()
 
+        uvloop.install()
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/app/ota_client_stub.py:90]: Only for performance improvement, if needed I can revert it and introduce it in a separate PR like the above one.

File: otaclient/ota_proxy/ota_cache.py [hunk 4]
```diff
@@ -112,29 +119,37 @@ def reader_on_done(self):
...
+
+        NOTE: must ensure cache entry is committed into the database
+        before calling this function!
         NOTE 2: this method is cross-thread method
+
+        Arg:
+            success: indicates whether the caching is successful or not.
         """
+        if not success:
+            logger.debug(f"writer failed on {self.fn}, abort...")
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/ota_proxy/ota_cache.py:133]: I will check the call chain and see if this log is needed.

File: otaclient/app/configs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (obi-t4) [reply otaclient/app/configs.py:97]: I think sometimes these configurations were changed in the previous PR w/o reason.
It is OK if the value is appropriate, but we are not sure about it  is appropriate or not if there is no reason/comment.

File: otaclient/app/ota_client_stub.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (obi-t4) [reply otaclient/app/ota_client_stub.py:90]: This time is OK, but next time you submit PR, to improve performance, another ticket and another PR is prefer.

File: otaclient/ota_proxy/ota_cache.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/ota_proxy/ota_cache.py:133]: Oh this line is not needed as the callers of this method all logs the reason of failure, I will remove this logging line.

File: otaclient/ota_proxy/config.py (hunk 1 above)
Comment (Bodong-Yang) [reply otaclient/ota_proxy/config.py]: Updated in [fad869d](https://github.com/tier4/ota-client/pull/183/commits/fad869d802f9e88bb07de73465a054a0941daf52).

File: tests/test_ota_proxy/test_ota_cache.py [hunk 5]
```diff
@@ -127,30 +127,30 @@ async def _wait_for_registeration_finish(self):
...
             return True, _tracker.meta  # type: ignore
-        else:
+        elif _tracker is not None:  # subscriber
             logger.debug(f"#{idx} is subscriber")
-            # wait for writter become ready
-            await _tracker.reader_subscribe(None)
             _tracker.reader_on_done()
             return False, _tracker.meta  # type: ignore
+        else:  # edge condition
+            return False, None
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply tests/test_ota_proxy/test_ota_cache.py]: Updated in [0e6a625](https://github.com/tier4/ota-client/pull/183/commits/0e6a62538eac2887d7ffc43a0299c954c8907df4).

File: otaclient/ota_proxy/ota_cache.py [hunk 6]
```diff
@@ -1047,16 +1016,18 @@ async def retrieve_file(
             raw_url, retry_cache=retry_cache
         ):
             logger.debug(f"use cache for {raw_url=}")
-            fp = _FileDescriptorHelper.open_cache(
-                meta_db_entry, base_dir=self._base_dir, executor=self._executor
+            fp = _FileDescriptorHelper.open_cached_file(
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/ota_proxy/ota_cache.py]: Updated in [0e6a625](https://github.com/tier4/ota-client/pull/183/commits/0e6a62538eac2887d7ffc43a0299c954c8907df4).

File: otaclient/ota_proxy/ota_cache.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/ota_proxy/ota_cache.py]: Updated in [0e6a625](https://github.com/tier4/ota-client/pull/183/commits/0e6a62538eac2887d7ffc43a0299c954c8907df4).

**Comment by Bodong-Yang** [comment]
@obi-t4 The PR reviews are resolved! Please have a look, thank you.

About the configuration changes and the introduce of `uvloop`, I add some explanation in the first PR comment.

**Comment by obi-t4** [comment]
@Bodong-Yang 

> This edge condition should be hard to trigger and can only result in one hash mismatched(retry one more time can resolve this issue and download the correct file). But in my local pytest running, this edge condition being triggered every time on the same file.

I'll create this ticket and fix it with another PR.
You may think that this issue doesn't affect on the real environment but I think it might have another potential issue on the background of this issue.
The cost of solving a issue on the actual production environment is much larger and harder than the development phase, so it should not be left as is.
Anyway this PR is OK.

**Comment by obi-t4** [comment]
@Bodong-Yang I'll start testing on the actual environment, thanks!

**Comment by obi-t4** [comment]
@Bodong-Yang 
It is working on the actual ecu!
If the previous comment is OK, this PR is OK.

**Comment by Bodong-Yang** [comment]
> 

Emmm I think multiple db threads is the cause of this problem, I think I need to set the db thread to 1 to resolve this problem.

**Comment by obi-t4** [comment]
> I think I need to set the db thread to 1 to resolve this problem.

This means, `sqlite3.OperationalError: database is locked` is unintended behavior?
If so, 
```
DB_THREAD_POOL_SIZE = 1
```
is not configurable, even it is in the configs.py.
Can you add comment that this value is not configurable.

And I'd like to know when this value is increased, what is improved?

**Review by obi-t4 — APPROVED** [review]
LGTM :+1:
